ocaml/typing/ctype.ml

```diff
@@ -2058,6 +2058,11 @@ let type_sort ~why env ty =
   | Ok _ -> Ok sort
   | Error _ as e -> e
 
+let type_sort_exn ~why env ty =
+  match type_sort ~why env ty with
```

Comment:
I don't understand this suggestion.  `Layout.sort_of_layout` wants a `layout` as an argument, but here we have a type.  Also, the current implementation should not succeed on `any` - that's the case where `type_sort` will return `Error`, so here you should get the `fatal_error`.  Note that, for example, the current requirement that functions have representable arguments and returns is implemented with calls to `type_sort`, so for example this program is correctly rejected:

```
type t_any [@@any]

type s = t_any -> int
```

I agree the check is guaranteed to succeed if callers respect the invariant in the comment.  That's why this is an `_exn` function.  I also think it's pretty cheap in situations where it succeeds - we can't need to load any new cmis, for example, because we already checked this type.

Reply:
My comment about `any` was too quick -- I meant that the current implementation would succeed on `'a : any`, but I think it shouldn't. And of course you're right about `sort_of_layout`. I suppose I had assumed there was an intervening call to `type_layout`, but that's slow.

I do agree that the common case in `constrain_type_layout` should be just about as quick as my alternative implementation. Let's leave this for now, but I may revisit this if/when we go to a more efficient way of getting from types to layouts.